### PR TITLE
feat: themer

### PR DESCRIPTION
**Pull request by nikosszzz**
* themer has been implemented, supports themes from Enmity aswell
* includes various fixes to Aliucord

### CODE REVIEW

File: src/api/Themer.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    applyTheme();
+}
+
+export function applyTheme() {
+    if (currentTheme === undefined) return;
+    setAMOLEDThemeEnabledBypass(false);
+    for (const key in Constants.ThemeColorMap) {
+        Constants.ThemeColorMap[key][2] = Constants.ThemeColorMap[key][0];
+        if (currentTheme.theme_color_map[key]) {
+            // 0 = Dark, 1 = Light, 2 = AMOLED
```
Comment (AAGaming00) [inline src/api/Themer.ts:24]: maybe an enum would be cleaner?

Comment (nikosszzz) [reply src/api/Themer.ts]: pushed enums for it

Comment (AAGaming00) [reply src/api/Themer.ts]: good

File: src/api/Themer.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import { Logger } from "../utils";
+import { THEME_DIRECTORY } from "../utils/constants";
+import { after } from "../utils/patcher";
+
+const logger = new Logger("Themer");
+export const themes = {} as Record<string, Theme>;
+export let currentTheme: Theme;
+
+enum ThemeType {
+    AMOLED = 2,
```
Comment (AAGaming00) [inline src/api/Themer.ts:13]: dumb just reverse order and don't specify value it automatically counts up from zero

explode

**Review by AAGaming00 — DISMISSED** [review]

File: src/api/Themer.ts (hunk 2 above)
Comment (nikosszzz) [reply src/api/Themer.ts]: done

**Comment by rushiiMachine** [comment]
Can we have themer toggleable completely, ie. it doesn't patch or load at all if it's disabled (pretty heavy patches after all)

File: src/api/Themer.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+export function setTheme(theme: Theme) {
+    window.Aliucord.settings.set("theme", theme.name);
+    currentTheme = theme;
+    applyTheme();
+}
+
+export function applyTheme() {
+    if (currentTheme === undefined) return;
```
Comment (rushiiMachine) [inline src/api/Themer.ts:25]: what about resetting theme? don't we need to store the original color map

**Comment by AAGaming00** [comment]
> Can we have themer toggleable completely, ie. it doesn't patch or load at all if it's disabled (pretty heavy patches after all)

yop do this

File: src/api/Themer.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    for (const file of readdir(THEME_DIRECTORY)) {
+        if (!file.name.endsWith(".json")) continue;
+
+        const themeFile = JSON.parse(AliuFS.readFile(THEME_DIRECTORY + file.name, "text") as string) as Theme;
+        themes[themeFile.name] = themeFile;
+    }
+
+    if (themes[window.Aliucord.settings.get("theme", "")] !== undefined) {
+        currentTheme = themes[window.Aliucord.settings.get("theme", "")];
```
Comment (rushiiMachine) [inline src/api/Themer.ts:63]: what's the point of this if/else?
current theme is going to be undefined anyways even if the theme setting is ""

**Review by rushiiMachine — CHANGES_REQUESTED** [review]

File: src/api/Themer.ts (hunk 4 above)
Comment (nikosszzz) [reply src/api/Themer.ts:63]: it checks if the theme file actually exists at the themes[...] part, this caused error when opening themer page without it

File: src/api/Themer.ts (hunk 3 above)
Comment (nikosszzz) [reply src/api/Themer.ts:25]: tried that before, even if i restore original color map discord just breaks its theme somehow

File: src/ui/patchTheme.ts [hunk 5]
```diff
@@ -1,35 +1,26 @@
...
-
         // 'I18N_LOAD_START' dispatch is the best time I can find to override the theme without breaking it.
         // Therefore, there's no guarantee that this will fix it for everyone
         if (ThemeStore) {
             const overrideTheme = () => {
                 ThemeManager.overrideTheme(ThemeStore.theme ?? "dark");
                 logger.info(`Overrode theme to ${ThemeStore.theme ?? "dark"}`);
+
+                if (AMOLEDThemeManager) {
+                    if (UnsyncedUserSettingsStore.useAMOLEDTheme === 2) {
```
[the remark this answers is not in the document]
Comment (nikosszzz) [reply src/ui/patchTheme.ts:16]: yop

File: src/core-plugins/CommandHandler.ts [hunk 6]
```diff
@@ -1,12 +1,17 @@
 import { ApplicationCommand, ApplicationCommandType, Commands } from "../api/Commands";
 import { Plugin } from "../entities/Plugin";
-import { getByProps } from "../metro";
+import { getByProps, SearchStore } from "../metro";
```
[the remark this answers is not in the document]
Comment (nikosszzz) [reply src/core-plugins/CommandHandler.ts:3]: guh is it necessary to do that

[the remark this answers is not in the document]
Comment (rushiiMachine) [reply src/core-plugins/CommandHandler.ts:3]: you shall do :trollface:

**Comment by AAGaming00** [comment]
this pr will have 9 million commits before it's merged

File: src/core-plugins/CommandHandler.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nikosszzz) [reply src/core-plugins/CommandHandler.ts:3]: guhhh alright

File: src/api/Themer.ts [hunk 7]
```diff
@@ -22,63 +24,75 @@ export function setTheme(theme: Theme) {
...
+            Constants.ThemeColorMap.CHAT_BACKGROUND[ThemeType.DARK] = currentTheme.theme_color_map["BACKGROUND_PRIMARY"][ThemeType.DARK];
+        }
 
-    setAMOLEDThemeEnabledBypass(true);
-    logger.info("Theme applied: ", currentTheme.name);
+        setAMOLEDThemeEnabledBypass(true);
+        logger.info("Theme applied: ", currentTheme.name);
+
+    } catch (e) {
+        logger.error("Failed to apply theme: ", e);
```
Comment (rushiiMachine) [inline src/api/Themer.ts:53]: error dialog?

**Review by rushiiMachine — CHANGES_REQUESTED** [review]

File: src/api/Themer.ts (hunk 7 above)
Comment (nikosszzz) [reply src/api/Themer.ts:70]: done

File: src/api/Themer.ts [hunk 8]
```diff
@@ -49,11 +49,30 @@ export function applyTheme() {
...
+        });
+
         logger.error("Failed to apply theme: ", e);
+        window.Aliucord.settings.set("theme", "");
         
-        unpatchInput();
-        unpatchNav();
+        // Unpatch components
+        unpatchInput;
+        unpatchNav;
```
Comment (rushiiMachine) [inline src/api/Themer.ts:67]: ? what is this supposed to do

Comment (nikosszzz) [reply src/api/Themer.ts:67]: hermes crashes the app if i put (), its the unpatch funcs

File: src/api/Themer.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+enum ThemeType {
+    DARK,
+    LIGHT,
+    AMOLED
+}
+
+const { externalStorageDirectory } = window.nativeModuleProxy.AliucordNative;
+const SETTINGS_DIRECTORY = externalStorageDirectory + "/AliucordRN/settings/";
+const THEME_DIRECTORY = externalStorageDirectory + "/AliucordRN/themes/";
+const settings = JSON.parse(AliuFS.readFile(SETTINGS_DIRECTORY + "Aliucord.json", "text") as string);
```
Comment (rushiiMachine) [inline src/api/Themer.ts:17]: shouldn't this fail terribly on first load

File: src/metro/index.ts [hunk 10]
```diff
@@ -44,6 +46,14 @@ for (const key in modules) {
         }
     }
 
+    if (!themeModuleFound && module?.publicModule?.exports?.ThemeColorMap) {
+        // Theme colors are overwritten here
+        themerInit(module.publicModule.exports);
+
+        themeModuleFound = true;
+        continue;
+    }
```
Comment (rushiiMachine) [inline src/metro/index.ts:55]: should you not error once you've looped over each and theme module still wasn't found?

File: src/ui/patchTheme.ts [hunk 11]
```diff
@@ -1,35 +1,42 @@
...
+import { currentTheme, themeApplied, themeErrorReason } from "../api/Themer";
+import { AMOLEDThemeManager, Dialog, FluxDispatcher, ReactNative, ThemeManager, ThemeStore, UnsyncedUserSettingsStore } from "../metro";
 
 export default function patchTheme() {
-    logger.info("Patching theme...");
+    if (!themeApplied && themeErrorReason) {
+        logger.error("Failed to apply theme: ", themeErrorReason);
+        Dialog.show({
+            title: "Failed to apply theme",
+            body: `${currentTheme?.name} failed to apply. Please report this issue. Theme will be disabled on restart.`,
```
Comment (rushiiMachine) [inline src/ui/patchTheme.ts:10]: > Please report this issue.

inb4 theme somehow breaks aliu and we will get 500 reports

File: src/api/Themer.ts [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+        for (const file of AliuFS.readdir(THEME_DIRECTORY)) {
+            if (!file.name.endsWith(".json")) continue;
+
+            const themeFile = JSON.parse(AliuFS.readFile(THEME_DIRECTORY + file.name, "text") as string) as Theme;
+
+            themes[themeFile.name] = themeFile;
+        }
+
+        if (settings !== undefined && themes[settings.theme] !== undefined) {
```
Comment (rushiiMachine) [inline src/api/Themer.ts:48]:
```suggestion
        if (themes[settings?.theme]) {
```

File: src/api/Themer.ts [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    try {
+        discordConstants = constants;
+        AliuHermes.unfreeze(constants.ThemeColorMap);
+        AliuHermes.unfreeze(constants.Colors);
+        AliuHermes.unfreeze(constants.UNSAFE_Colors);
+
+        for (const file of AliuFS.readdir(THEME_DIRECTORY)) {
+            if (!file.name.endsWith(".json")) continue;
+
+            const themeFile = JSON.parse(AliuFS.readFile(THEME_DIRECTORY + file.name, "text") as string) as Theme;
```
Comment (rushiiMachine) [inline src/api/Themer.ts:43]: should check for validity (esp. name not missing/empty)

File: src/api/Themer.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+            themes[themeFile.name] = themeFile;
```
Comment (rushiiMachine) [inline src/api/Themer.ts:45]: check for themes with duplicate names?

File: src/api/Themer.ts [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            discordConstants.ThemeColorMap[key][ThemeType.AMOLED] = discordConstants.ThemeColorMap[key][ThemeType.DARK];
+            if (currentTheme.theme_color_map[key]) {
+                discordConstants.ThemeColorMap[key][ThemeType.AMOLED] = currentTheme.theme_color_map[key]?.[ThemeType.DARK];
+                discordConstants.ThemeColorMap[key][ThemeType.LIGHT] = currentTheme.theme_color_map[key]?.[ThemeType.LIGHT];
+                discordConstants.ThemeColorMap[key][ThemeType.DARK] = currentTheme.theme_color_map[key]?.[ThemeType.DARK];
+            }
+        }
+
+        // Enmity compat
+        if (currentTheme.colors && currentTheme.colours === undefined) {
```
Comment (rushiiMachine) [inline src/api/Themer.ts:77]:
```suggestion
        if (currentTheme.colors && !currentTheme.colours) {
```

**Review by rushiiMachine — CHANGES_REQUESTED** [review]

File: src/api/Themer.ts (hunk 9 above)
Comment (nikosszzz) [reply src/api/Themer.ts:17]: nop it works fine

Comment (nikosszzz) [reply src/api/Themer.ts:17]: well actually yes wait

File: src/api/Themer.ts [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+            themes[themeFile.name] = themeFile;
+        }
+
+        if (themes[settings?.theme]) {
+            currentTheme = themes[settings.theme];
+
+            applyTheme();
+        } else {
+            return;
+        }
```
Comment (janisslsm) [inline src/api/Themer.ts:59]: Useless return statement cause no further code gets executed :trolleybus:

Comment (nikosszzz) [reply src/api/Themer.ts:59]: ill remove trolley

**Review by janisslsm — APPROVED** [review]
lgtm

File: src/api/Themer.ts [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+// WARNING: this function is called before Aliucord loads, meaning we're having limited access.
+export function applyTheme() {
+    try {
+        if (currentTheme === undefined) return;
+        for (const key in currentTheme.theme_color_map) {
+            if (!discordConstants.ThemeColorMap[key]) continue;
+
+            discordConstants.ThemeColorMap[key][ThemeType.AMOLED] = discordConstants.ThemeColorMap[key][ThemeType.DARK];
+            if (currentTheme.theme_color_map[key]) {
+                discordConstants.ThemeColorMap[key][ThemeType.AMOLED] = currentTheme.theme_color_map[key]?.[ThemeType.DARK];
```
Comment (amsyarasyiq) [inline src/api/Themer.ts:73]: We don't have to force the user to use AMOLED anymore, hence just overwrite the regular dark theme fields.

File: src/api/Themer.ts [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        for (const key in currentTheme.unsafe_colors) {
+            if (!discordConstants.UNSAFE_Colors[key]) continue;
+
+            discordConstants.UNSAFE_Colors[key] = currentTheme.unsafe_colors[key];
+        }
+
+        if (!currentTheme.theme_color_map["CHAT_BACKGROUND"]) {
+            discordConstants.ThemeColorMap.CHAT_BACKGROUND[ThemeType.AMOLED] = currentTheme.theme_color_map["BACKGROUND_PRIMARY"][ThemeType.DARK];
```
Comment (amsyarasyiq) [inline src/api/Themer.ts:101]: This one as well.

**Review by rushiiMachine — APPROVED** [review]

File: src/api/Themer.ts [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+
+        themeApplied = true;
+    } catch (e) {
+        themeApplied = false;
+        themeErrorReason = e;
+    }
+}
+
+export function useDiscordThemes() {
+    window.Aliucord.settings.set("theme", "");
```
Comment (amsyarasyiq) [inline src/api/Themer.ts:114]: Any reason to not use `settings.delete`?

File: src/api/Themer.ts [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+
+            const themeFile = JSON.parse(AliuFS.readFile(THEMES_DIRECTORY + file.name, "text") as string) as Theme;
+
+            if (!themeFile?.name || !themeFile?.theme_color_map || !themeFile?.colors) throw new Error(`Theme file ${file.name} does not contain a name, theme_color_map or colors key.`);
+            if (themes[themeFile.name]) throw new Error(`A theme called ${themeFile.name} already exists.`);
+            
+            themes[themeFile.name] = themeFile;
+        }
+
+        if (themes[settings?.theme]) {
```
Comment (amsyarasyiq) [inline src/api/Themer.ts:53]:
```suggestion
        if (settings?.theme && themes[settings.theme]) {
```

File: src/metro/index.ts [hunk 21]
```diff
@@ -270,6 +286,16 @@ export const Clipboard = getByProps("getString", "setString") as {
     setString(str: string): Promise<void>;
 };
 
+export enum AMOLEDThemeState {
+    HIDDEN,
+    OFF,
+    ON
+}
+
+export const setAMOLEDThemeEnabledBypass = (state) => {
```
Comment (amsyarasyiq) [inline src/metro/index.ts:295]: As of latest commit, isn't this now unnecessary?

File: src/ui/patchTheme.ts [hunk 22]
```diff
@@ -1,35 +1,42 @@
...
-            }
-
-            patcher.before(AMOLEDThemeManager, "setAMOLEDThemeEnabled", (_, enabled) => {
-                logger.info(`Setting AMOLED theme to ${enabled}`);
-                window.Aliucord.settings.set("enableAMOLEDTheme", enabled);
-            });
-        } else {
-            logger.error("Could not get AMOLEDThemeManager");
-        }
+        window.Aliucord.settings.set("theme", "");
```
Comment (amsyarasyiq) [inline src/ui/patchTheme.ts:17]: Same with this one.

File: src/Aliucord.ts [hunk 23]
```diff
@@ -14,6 +14,7 @@ interface SettingsSchema {
     disablePluginsOnCrash: boolean;
     plugins: Record<string, boolean>;
     enableAMOLEDTheme: boolean;
```
Comment (amsyarasyiq) [inline src/Aliucord.ts:16]: This is no longer necessary.

**Review by amsyarasyiq — CHANGES_REQUESTED** [review]

**Pull request merged**